pfcpiface/config.go

```diff
@@ -0,0 +1,189 @@
+	PBS                uint32 `json:"pbs"`
+	EBS                uint32 `json:"ebs"`
+	BurstDurationMs    uint32 `json:"burst_duration_ms"`
+	SchedulingPriority uint32 `json:"priority"`
+}
+
+type SliceMeterConfig struct {
+	N6RateBps    uint64 `json:"n6_bps"`
+	N6BurstBytes uint64 `json:"n6_burst_bytes"`
+	N3RateBps    uint64 `json:"n3_bps"`
+	N3BurstBytes uint64 `json:"n3_burst_bytes"`
+}
+
+// SimModeInfo : Sim mode attributes.
+type SimModeInfo struct {
+	MaxSessions uint32 `json:"max_sessions"`
+	StartUEIP   net.IP `json:"start_ue_ip"`
+	StartENBIP  net.IP `json:"start_enb_ip"`
+	StartAUPFIP net.IP `json:"start_aupf_ip"`
+	N6AppIP     net.IP `json:"n6_app_ip"`
+	N9AppIP     net.IP `json:"n9_app_ip"`
+	StartN3TEID string `json:"start_n3_teid"`
+	StartN9TEID string `json:"start_n9_teid"`
+}
+
+// CPIfaceInfo : CPIface interface settings.
+type CPIfaceInfo struct {
+	Peers           []string `json:"peers"`
+	UseFQDN         bool     `json:"use_fqdn"`
+	NodeID          string   `json:"hostname"`
+	EnableUeIPAlloc bool     `json:"enable_ue_ip_alloc"`
+	UEIPPool        string   `json:"ue_ip_pool"`
+	HTTPPort        string   `json:"http_port"`
+	Dnn             string   `json:"dnn"`
+}
+
+// IfaceType : Gateway interface struct.
+type IfaceType struct {
+	IfName string `json:"ifname"`
+}
+
+// validateConf checks that the given config reaches a baseline of correctness.
+func validateConf(conf Conf) error {
+	validModes := map[string]struct{}{
+		"af_xdp":    {},
+		"af_packet": {},
+		"dpdk":      {},
+		"sim":       {},
+	}
+	if _, ok := validModes[conf.Mode]; !ok {
+		return ErrInvalidArgumentWithReason("conf.Mode", conf.Mode, "invalid mode")
+	}
+
+	if conf.EnableP4rt {
+		_, _, err := net.ParseCIDR(conf.P4rtcIface.AccessIP)
+		if err != nil {
+			return err
+		}
+		_, _, err = net.ParseCIDR(conf.P4rtcIface.UEIP)
+		if err != nil {
+			return err
+		}
+	}
+
+	if conf.CPIface.EnableUeIPAlloc {
+		_, _, err := net.ParseCIDR(conf.CPIface.UEIPPool)
+		if err != nil {
+			return err
+		}
+	}
+	for _, peer := range conf.CPIface.Peers {
+		ip := net.ParseIP(peer)
+		if ip != nil {
+			return ErrInvalidArgumentWithReason("conf.CPIface.Peers", peer, "invalid IP")
+		}
+	}
+
+	if _, err := log.ParseLevel(conf.LogLevel); err != nil {
+		return err
+	}
+
+	if _, err := time.ParseDuration(conf.RespTimeout); err != nil {
+		return ErrInvalidArgumentWithReason("conf.RespTimeout", conf.RespTimeout, "invalid duration")
+	}
+
+	if conf.ReadTimeout == 0 {
+		return ErrInvalidArgumentWithReason("conf.ReadTimeout", conf.ReadTimeout, "invalid duration")
+	}
+
+	if conf.MaxReqRetries == 0 {
+		return ErrInvalidArgumentWithReason("conf.MaxReqRetries", conf.MaxReqRetries, "invalid number of retries")
+	}
+
+	if conf.EnableHBTimer {
+		if _, err := time.ParseDuration(conf.HeartBeatInterval); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// ParseJSON : parse json file and populate corresponding struct.
+func ParseJSON(filepath string) (Conf, error) {
```

Comment:
```suggestion
func ParseConf(filepath string) (Conf, error) {
```

Reply:
I went with `LoadConfigFile`. Parse might mean just "parse json", but there's also file IO and validation.